Rt11xx DMA Support

danieldegrasse opened the pull request:
Enables DMA support for RT11xx SOCs and evaluation boards. DMA support is enabled on the Cortex M7 core, and has been tested using the `loop_transfer` dma test.

danieldegrasse committed:
soc: MIMXRT11xx update clock init code

RT11xx clock init code was not handling difference between RT1160 and
RT1170 Bus_Lpsr clock correctly. Fix this.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
boards: mimxrt1060_evk_cm7: Enable DMA controller on RT1160 EVK

DMA controller is enabled and tested for cortex M7 core. The M7 core was
tested on the loop and channel link transfer tests.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
boards: mimxrt1070_evk_cm7: Enable DMA controller on RT1170 EVK

DMA controller is enabled and tested for cortex M7 core. The M7 core was
tested on the loop and channel link transfer tests.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

mbolivar-nordic commented on include/arch/arc/v2/linker.ld:
```diff
@@ -26,6 +26,12 @@
 		#define RAMABLE_REGION SRAM
 	#endif
 #endif
+/* Non cacheable memory region, can be overridden by SOC */
```
@tejlmand is this a job for snippets?

tejlmand replied:
The `nocache.ld` is already being included as a snippet.
https://github.com/zephyrproject-rtos/zephyr/blob/2be4a537afa19878533909cea95d3fb96f6ffa98/arch/common/CMakeLists.txt#L41-L44

so instead of adding those `#ifdef` to all linker.ld files, then we could adjust the snippet instead, like https://github.com/zephyrproject-rtos/zephyr/blob/main/arch/common/nocache.ld:
```
#ifdef CONFIG_ARCH_HAS_NOCACHE_MEMORY_SUPPORT
} GROUP_DATA_LINK_IN(NOCACHE_REGION, NOCACHE_REGION)
#else
} GROUP_DATA_LINK_IN(RAMABLE_REGION, RAMABLE_REGION)
#endif
```

One concern I have with this, is the fact that this snippet is still injected in the ram section here:
https://github.com/zephyrproject-rtos/zephyr/blob/2be4a537afa19878533909cea95d3fb96f6ffa98/include/arch/arm/aarch32/cortex_m/scripts/linker.ld#L251-L254

So we should consider adding an extra nocache snippet here:
https://github.com/zephyrproject-rtos/zephyr/blob/2be4a537afa19878533909cea95d3fb96f6ffa98/cmake/extensions.cmake#L1170-L1179

that are then used when `CONFIG_ARCH_HAS_NOCACHE_MEMORY_SUPPORT=y`.

carlocaione replied:
Good time for a bit of promotion to https://github.com/zephyrproject-rtos/zephyr/issues/40146

danieldegrasse replied:
> so instead of adding those #ifdef to all linker.ld files, then we could adjust the snippet instead, like 
> ```
> #ifdef CONFIG_ARCH_HAS_NOCACHE_MEMORY_SUPPORT
> } GROUP_DATA_LINK_IN(NOCACHE_REGION, NOCACHE_REGION)
> #else
> } GROUP_DATA_LINK_IN(RAMABLE_REGION, RAMABLE_REGION)
> #endif
> ```

@tejlmand Just to make sure I'm understanding this snippet correctly, in the case of the Cortex M4 would the `NOCACHE_REGION` be defined by the SOC linker file? In that case, I think that a new KConfig symbol needs to be defined, as SOCs with a Cortex M7 also select `CONFIG_ARCH_HAS_NOCACHE_MEMORY_SUPPORT`, but do not define a `NOCACHE_REGION`, correct?

tejlmand replied:
> @tejlmand Just to make sure I'm understanding this snippet correctly, in the case of the Cortex M4 would the NOCACHE_REGION be defined by the SOC linker file? 

Something along that was my though.
The soc linker file or the general arch linker script, where it could be depending on the devicetree information.
Similar to the SRAMx regions, and the description of #40146.

> In that case, I think that a new KConfig symbol needs to be defined, as SOCs with a Cortex M7 also select CONFIG_ARCH_HAS_NOCACHE_MEMORY_SUPPORT, but do not define a NOCACHE_REGION, correct?

Good point.
So some thinking must be done if going into this direction.

danieldegrasse commented:
I've updated this PR to remove the CM4 support until a well defined system exists within Zephyr to define a custom `nocache` region, and removed all the changes to the linker files. The PR now supports only the CM7 core on the RT1170/1160, but has a much smaller changeset.

danieldegrasse committed:
drivers: clock_control: Add LPSR clock for EDMA

Adds ccm clock config for EDMA1 clock

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
drivers: edma: update MCUX dma driver for multi-instance

Updates mcux edma driver to handle multiple DMA instances correctly.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse commented:
@hakehuang Integrated changes from #40828:
- DMA driver supports multiple instances
- DMA1 LPSR clock source defined in clock management driver

hakehuang commented on drivers/clock_control/clock_control_mcux_ccm_rev2.c:
```diff
@@ -78,6 +81,7 @@ static int mcux_ccm_get_subsys_rate(const struct device *dev,
 		clock_root = kCLOCK_Root_Gpt1 + instance;
 		break;
 #endif
+
```
remove empty line

hakehuang commented on dts/arm/nxp/nxp_rt1160_cm4.dtsi:
```diff
@@ -60,10 +60,10 @@
 			clocks = <&ccm IMX_CCM_EDMA_CLK 0x7C 0x000000C0>;
 			status = "disabled";
-			label = "EDMA1";
+			label = "DMA_1";
```
this coould be DMA_0, as for m4 core this is the only dma, which will aligned with test application default setttings

hakehuang commented:
@danieldegrasse ,  for dts there is one more changed need
```
dma-requests = <208>;
```

hakehuang commented on dts/arm/nxp/nxp_rt1170_cm4.dtsi:
```diff
@@ -60,10 +60,10 @@
 			clocks = <&ccm IMX_CCM_EDMA_CLK 0x7C 0x000000C0>;
 			status = "disabled";
-			label = "EDMA1";
+			label = "DMA_1";
```
same as above

danieldegrasse commented:
@hakehuang I've implemented the requested changes, please let me know if you see anything else missing

hakehuang commented:
OK for me, but need resolve the conflicts

cfriedt commented:
@danieldegrasse - please rebase and fix conflicts

danieldegrasse commented:
@cfriedt- Fixed, should be good to go
